### DISCUSSION TITLE
Limitations and weirdness of Outpost generation

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#8615] [Unstable feedback]
So I've been trying to experiment a little bit with colonies with my limited knowledge of outpost generation, in an attempt to spice up the very strict layout of this type of outposts, and maybe improving outposts in general.
Unfortunately I kept hitting a significant amount of hurdles caused by the somewhat rigid outpost generation logic.

The result is very, very cursed: _Almost everything goes wrong in this one_
![image](https://user-images.githubusercontent.com/73229309/160145369-91829b73-d8bd-4812-86a2-8261c3c0ce3e.png)

You can find my experiment here:
https://steamcommunity.com/sharedfiles/filedetails/?id=2783676590

So here's a quick list of things that feel... broken about outpost generation

**1. Stairs > Ladders**

The number one problem with outposts is **ladders**. 
Ideally, outposts should not use ladders and hatches between rooms. Unfortunately, it's much easier for outpost generation to use ladders over stairs. The vertical hallway can stretch a ladder infinitely. 

This is something that should be revised, as there needs to be a way to smoothly connect modules via stairs, which implies stacking "stairway" vertical modules (which can definitely happen as these are 3x the same stair module).
![image](https://user-images.githubusercontent.com/73229309/160141688-1f017252-01ca-42dd-a81f-3c268c9147ea.png)

**2. Possibility of "no hallways"**

You cannot prevent the default Horizontal/Vertical stretchy hallways from being used. Outpost generation will liberally create huge Hallways rather than using a different "connector" piece. 

![image](https://user-images.githubusercontent.com/73229309/160142349-c681c75f-0e6a-437d-b1ad-0ebe72b87c88.png)

**3. Defining "not dead end"**

Some modules have only one entrance and can appear at the edge of the outpost. These are by all means dead-ends. But other modules with two or more entrances can also be generated at the edge of the map unnecessarily.
![image](https://user-images.githubusercontent.com/73229309/160139399-1e5b9735-0008-40e4-a4cc-0d4169de39cf.png)
Hallways sometimes generated as a dead end
![image](https://user-images.githubusercontent.com/73229309/160139942-f13e2070-d8f9-4cfb-9746-d639d978b571.png)

There could be an option to define a module as "not a dead end" so it does not unnecessarily spawn at the edges of the outpost. 
This is to avoid mostly empty "connector" modules like hallways to spawn at the edges. Ruins could also benefit from this feature:
![image](https://user-images.githubusercontent.com/73229309/160605330-36a2e96e-482a-4a88-9d2f-1e73178d4d8b.png)

Example of a connector/decorative module that spawns at the edge of an outpost
![image](https://user-images.githubusercontent.com/73229309/160142894-c5237b98-f66a-4373-9e4a-d3759c40478d.png)

**4. Floating amount of modules**

If you are trying to use certain modules as hallway substitutes, you HAVE to define a finite amount of them. 
There should be a way to not use a define amount, but instead leave the outpost generation logic handle it by adding the necessary amount of modules according to allowed module connections. 

For more variation, a fork would also be useful if you are trying to generate a certain amount of "decoration" or "filler" modules, or not always spawn some module types. 

**5. Refusing to spawn modules**

There are very clear cases of the outpost generation simply refusing to spawn modules that could otherwise be easily placed, but don't. 

This one is lacking some modules that have a clear connection potential on their left, but won't. 
![image](https://user-images.githubusercontent.com/73229309/160147316-3d282ec5-f617-4616-bd57-38b34ed3f65f.png)

**6. Can't accomodate stairs**
#8606

**7. What does "CanAttachToPrevious" do?** 

It's unclear what it does, or if it does anything. 
Some modules check all 4 directions despite having only 2 gaps.

**8. "Allow attaching to" is not respected?**

Some modules only have one attachment type but still attach to others.
Example: this module only attaches to "Airlock", but it is connected to the Residential module on its right
![image](https://user-images.githubusercontent.com/73229309/160183092-3cd8ffff-a1b0-43e8-93cd-9e8a9621cd07.png)

**Comment by Regalis11** [comment]
>2. You cannot prevent the default Horizontal/Vertical stretchy hallways from being used. Outpost generation will liberally create huge Hallways rather than using a different "connector" piece.

You can be designing the modules and the layout in a way that doesn't require offsetting the modules from each other. The hallways are only used if the modules need to be moved further from each other to prevent them from overlapping. And as you yourself mentioned in the previous point, it is possible to prevent the vertical hallways from being used by using stairs instead of ladders.

> What does "CanAttachToPrevious" do?

"Which sides of this module are allowed to attach to the previously placed module. E.g. if you want a module to always attach to the left side of the docking module, you could set this to Right."

> If you are trying to use certain modules as hallway substitutes, you HAVE to define a finite amount of them.

I assume you're trying to spawn these "hallway substitutes" by forcing the outpost to include a certain number of them? I think it'd work better if you just increased the total number of modules and let the generator choose the "unspecified" extra modules from the pool of those hallway substitutes.

**Reply by ubertpendragon** [reply]
> You can be designing the modules and the layout in a way that doesn't require offsetting the modules from each other. The hallways are only used if the modules need to be moved further from each other to prevent them from overlapping. 

As you can see it *will* sometimes generate hallways even if there's no chance of overlapping
![image](https://user-images.githubusercontent.com/73229309/160406329-3bdd1dcf-ad37-44ca-a26e-5bed07a6297a.png)

I'm not sure what makes it *refuse* to overlap things.

Sometimes the hulls will overlap
![image](https://user-images.githubusercontent.com/73229309/160411946-a98b61bc-7108-45b6-a58f-167d4630e2b3.png)

Sometimes the walls will overlap
![image](https://user-images.githubusercontent.com/73229309/160412117-0f230cec-10bc-4749-b0c3-73cc3285c10d.png)

And then the game will create an offset that probably shouldn't exist: 
![image](https://user-images.githubusercontent.com/73229309/160412469-bff3c788-05e3-47a4-aae9-d33dc7c769a7.png)

The way current colonies work without having to generate hallways is their extremely rigid layout. I hoped that the editor would try to connect modules with as little hallways as possible but there's another logic that conflicts "simpler" connections.


> And as you yourself mentioned in the previous point, it is possible to prevent the vertical hallways from being used by using stairs instead of ladders.

I'll try this (when the save button gets fixed)

> What does "CanAttachToPrevious" do?
> 
> "Which sides of this module are allowed to attach to the previously placed module. E.g. if you want a module to always attach to the left side of the docking module, you could set this to Right."

Oh yes sorry wrong screencap. I just don't understand the wording and/or logic on this. "Previously", and module pieces that only have L/R connections somehow having all directions checked. 
Do you set "right" on the docking module? Assuming "previously" what's the order of modules being placed? Are modules generated around the docking module?
Maybe I'm a little daft but since so few people have been doing outpost editing it's hard to find a clearer explanation


> I assume you're trying to spawn these "hallway substitutes" by forcing the outpost to include a certain number of them?

I've been using modules with the `AOG_colonyhallway` tag, and been cranking their numbers up to generate a lot of them (which the outpost generator refuses to do) in the hopes of getting proper connections. Ideally I'd let the editor decide

> I think it'd work better if you just increased the total number of modules and let the generator choose the "unspecified" extra modules from the pool of those hallway substitutes.

I'm not sure how to do that. I've got a total of 8 different modules with all sorts of connection configurations among the `AOG_colonyhallway` pool. By all means, even at its worst, outpost generation should always be able to prevent error messages and conflicts by turning the outpost into a thin long stretch of left/right connected rooms, but it doesn't.
